rtbkit/plugins/augmentor_interface/http_augmentor_interface.cc

```diff
@@ -0,0 +1,486 @@
+            loop.addSource("HttpAugmentorInterface::httpClientAugmentor",
+                                inst.httpClientAugmentor);
+
+            loop.addPeriodic("HttpAugmentorInterface::reportQueues", 1.0, [=](uint64_t) {
+                recordLevel(inst.httpClientAugmentor->queuedRequests(), "queuedRequests");
+            });
+
+            AugmentorInfoEntry e;
+            e.name = augName;
+
+            AugmentorInfoMap::iterator it;
+            if((it = augmentors.find(augName)) == augmentors.end()){
+                std::shared_ptr<AugmentorInfo> info = std::make_shared<AugmentorInfo>(augName);
+                info->instances.push_back(inst);
+                augmentors.insert(make_pair(augName, info));
+                std::cerr << "inserting : " << augName << std::endl;
+                e.info = info;
+            }else{
+                it->second->instances.push_back(inst);
+                e.info = it->second;
+            }
+            allAugmentors->push_back(e);
+        }
+    } catch (const std::exception & e) {
+        THROW(error) << "configuration file is invalid" << std::endl
+            << "usage : " << std::endl
+            << "{" << std::endl
+            << "\t\"type\": \"http\"," << std::endl
+            << "\t\"augmentors\": [" << std::endl
+            << "\t\t{" << std::endl
+            << "\t\t\t\"name\": \"augmentor01\"," << std::endl
+            << "\t\t\t\"host\": \"http://localhost\"," << std::endl
+            << "\t\t\t\"path\": \"/augmentor01\"" << std::endl
+            << "\t\t\t\"maxInFlight\": 1024" << std::endl
+            << "\t\t}," << std::endl
+            << "\t\t{" << std::endl
+            << "\t\t\t\"name\": \"augmentor02\"," << std::endl
+            << "\t\t\t\"host\": \"http://localhost\"," << std::endl
+            << "\t\t\t\"path\": \"/augmentor02\"" << std::endl
+            << "\t\t\t\"maxInFlight\": 1024" << std::endl
+            << "\t\t}"<< std::endl
+            << "\t]" << std::endl
+            << "}" << std::endl;
+    }
+    
+}
+
+HttpAugmentorInterface::~HttpAugmentorInterface()
+{
+    shutdown();
+}
+
+void HttpAugmentorInterface::init(Router * r){
+    router = r;
+ 
+    inbox.onEvent = [&] (const std::shared_ptr<Entry>& entry)
+        {
+            doAugmentation(entry);
+        };
+
+    loop.addSource("AugmentationLoop::inbox", inbox);
+    
+    loop.addPeriodic("AugmentationLoop::checkExpiries", 0.001,
+                [=] (int) { checkExpiries(); });
+
+    loop.addPeriodic("AugmentationLoop::recordStats", 0.977,
+                [=] (int) { recordStats(); });
+}
+
+void HttpAugmentorInterface::start() {
+    loop.start();
+}
+
+void HttpAugmentorInterface::sleepUntilIdle()
+{
+    while (!idle_)
+        ML::futex_wait(idle_, 0);
+}
+
+void HttpAugmentorInterface::shutdown() {
+    loop.shutdown();
+}
+
+size_t HttpAugmentorInterface::numAugmenting() const {
+    return augmenting.size();
+}
+
+void HttpAugmentorInterface::augment(
```

Comment:
I'm not sure to understand why you did not keep this logic in the AugmentationLoop itself ?